File: src/leetcode/python/DecodeString_394.py
class DecodeString_394(object):
    def decodeString(self, s):
        """
        :type s: str
        :rtype: str
        """
        stack = [["", 1]]
        num = ""
        for c in s:
            if c.isdigit():
                num += c
            elif c == '[':
                stack.append(["", int(num)])
                num = ""
            elif c == ']':
                seq, fre = stack.pop()
                stack[-1][0] += seq * fre
            else:
                stack[-1][0] += c
        return stack[0][0]

File: src/leetcode/java/DecodeString_394.java
package leetcode.java;

import java.util.Stack;

public class DecodeString_394 {
  public String decodeString(String s) {
    Stack<Integer> fre = new Stack<>();
    Stack<String> res = new Stack<>();
    res.push("");
    for (int i = 0; i < s.length(); i++) {
      if ('0' <= s.charAt(i) && s.charAt(i) <= '9') {
        int j = i;
        while ('0' <= s.charAt(i + 1) && s.charAt(i + 1) <= '9') i++;
        fre.push(Integer.valueOf(s.substring(j, i + 1)));
      } else if (s.charAt(i) == '[') {
        res.push("");
      } else if (s.charAt(i) == ']') {
        int freq = fre.pop();
        String sub = res.pop();
        StringBuilder bd = new StringBuilder();
        for (int j = 0; j < freq; j++) bd.append(sub);
        res.push(res.pop() + bd.toString());
      } else {
        res.push(res.pop() + s.charAt(i));
      }
    }
    return res.pop();
  }
}
